const getContextObject = require('../utils/context-object')
const { parseDeleteUrl } = require('../utils/parse-url')
const handleError = require('../utils/handle-error')

const del = service => {
  return (req, res) => {
    let parsedUrl
    try {
      parsedUrl = parseDeleteUrl(service.entities, req)
    } catch (err) {
      return handleError(err, service, res)
    }

    const context = getContextObject(service, parsedUrl, req, res)

    return service
      .processEvent('DELETE', context)
      .then(() => {
        res.status(204).send()
      })
      .catch(err => {
        // Hide errors in generic message but log detailed error
        handleError(err, service, res)
      })
  }
}

module.exports = del
